Minor bugfix: Check retval for termination by signal

nzalev opened the pull request:
This modifies `get_retval`. Under normal circumstances, the exit code is stored in the upper 8 bits of the 16 bit status variable, so it is correct to divide by 256.

However, if Rsync is terminated by a signal, it does not have a normal return code. Instead, the terminating signal is stored in the lower 7 bits and the expected return code (given by a shell) is 128 plus the signal number.

There are some cases where Rsync can be terminated by SIGSEGV, in which case the expected return code is 139. Instead, a floating point value is currently outputted:

```
ERROR: /usr/local/bin/rsync returned 0.54296875 while processing root@my.host.net:/
```

This patch fixes this by checking the equivalent of `WIFSIGNALED` and will instead return 128 + signo.

nzalev commented:
Doesn't look like there will be any additional input ;)

sgpinkus commented:
Will test today :).

sgpinkus commented:
There is definitely an issue here. Goes back to much earlier versions. However, I'm not sure this fix is quite right. Test conf:

```
config_version	1.2
snapshot_root	/tmp/test-backup-dest/
lockfile	/tmp/rsnapshot-test.lock
logfile	/tmp/rsnapshot.log
cmd_rsync	/usr/bin/rsync
retain		alpha	2
sync_first	1
cmd_preexec	/usr/bin/sleep infinity
backup	/tmp/test-backup-src/		localhost/
```


Then run and `Ctrl+C` (SIGINT) it:

```
$ ~/src/rsnapshot/rsnapshot -c rsnapshot.conf sync
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
/home/sam/src/rsnapshot/rsnapshot -c rsnapshot.conf sync 
----------------------------------------------------------------------------
ERROR: cmd_preexec "/usr/bin/sleep infinity" returned 0.0078125
```

Reading the [doc](https://perldoc.perl.org/functions/system) seems like the correct handling of return value from `system` should be this:

```
diff --git a/rsnapshot-program.pl b/rsnapshot-program.pl
index 518af81..6bcbd6d 100755
--- a/rsnapshot-program.pl
+++ b/rsnapshot-program.pl
@@ -6090,8 +6090,10 @@ sub get_retval {
        if ($retval !~ m/^\d+$/) {
                bail("get_retval() was passed $retval, a number is required");
        }
-
-       return ($retval / 256);
+       if ($retval & 127) {
+               return $retval;
+       }
+       return $retval >> 8;
 }
```

sgpinkus reviewed (changes requested):
See comment.

nzalev commented:
Apologies, the missing parenthesis in `if ($retval & 0x7f > 0)` cause it to be ineffective.
Changing this to `if ($retval & 0x7f)` or enclosing the bitwise and in parenthesis fixes this. 

However, in your diff, you are only returning the signal number which is unusual. With `return ($retval & 127);`, any signal which cannot be handled by Rsync has its signal number mixed in with its regular return codes.  For example:
```
./rsnapshot -c test.conf sync
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
./rsnapshot -c test.conf sync 
----------------------------------------------------------------------------
ERROR: /usr/bin/rsync returned 11 while processing /tmp/test-backup-src/
```

However, Rsync did not exit due to an error in file IO, which it normally represents with exit code 11. Instead, it died to signal number 11, which are two different situations. This is _typically_ handled by returning 128 plus the signal number. Take the following examples from different shells and different OSs:
```
noah@E945:~/wrk/rsnapshot$ uname; echo $SHELL; ./segfault; echo $?
Linux
/bin/bash
Segmentation fault (core dumped)
139


:winpkg01:/home/user/noahzale> uname; echo $SHELL; ./segfault; echo $?
AIX
/usr/bin/ksh
Segmentation fault(coredump)
139


meshc2# uname; echo $SHELL; ./segfault; echo $?
OpenBSD
/bin/sh
Segmentation fault (core dumped) 
139


# similarly without a core dump
noah@E945:~$ cat
^C
noah@E945:~$ echo $?
130
```



In my original comment, the example I gave is from a real production system:
```
ERROR: /usr/local/bin/rsync returned 0.54296875 while processing root@my.host.net:/
```

If you multiply out 0.54296875 by 256 (as  `$retval` is divided by 256 in the current function), you get back the original status value of 139. This is signal 11 with the 8th bit (core dump flag) set, which is a legitimate segmentation fault, so my examples are not so contrived ;).


---

So, I will fix the issue with the parenthesis, and I can change 0x7f to 127 if you prefer as well. I will wait though for feedback on which way you prefer to return the signal bits.

For the regular return value, I personally prefer bit shifting by 8 as opposed to dividing by 256. But as long as the signal and core dump bits are properly handled, there is no functional difference, so it is just stylistic. For that reason, I didn't touch it, but I can change it if you like.

sgpinkus commented:
> However, in your diff, you are only returning the signal number which is unusual. With return ($retval & 127);

Yes, your right, `$retval` should not be masked. Should've been:

```
diff --git a/rsnapshot-program.pl b/rsnapshot-program.pl
index 518af81..6bcbd6d 100755
--- a/rsnapshot-program.pl
+++ b/rsnapshot-program.pl
@@ -6090,8 +6090,10 @@ sub get_retval {
        if ($retval !~ m/^\d+$/) {
                bail("get_retval() was passed $retval, a number is required");
        }
-
-       return ($retval / 256);
+       if ($retval & 127) {
+               return $retval;
+       }
+       return $retval >> 8;
 }
```

I prefer `>> 8`  and have no preference for 0x7f and 127. Will accept either but whatever is consistent with the rest of the code base is better - if relevant.

nzalev commented:
With respect to returning `$retval` directly, that is in some sense worse than returning just the signal number.

The 8th bit in `$retval` is used as a status flag to indicate whether a core dump would have occurred. It is sometimes set as 1 and sometimes not, which creates a somewhat inconsistent behaviour with the return codes.

By returning `$retval` directly, the return code is sometimes just the signal number and is sometimes 128 plus the signal number in cases where a core dump would have occurred. Here is an example. The signals used here are SIGINT (2) and SIGSEGV (11).
```
./rsnapshot -c test.conf sync
^C----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
./rsnapshot -c test.conf sync 
----------------------------------------------------------------------------
ERROR: cmd_preexec "/usr/bin/sleep infinity" returned 2



./rsnapshot -c test.conf sync
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
./rsnapshot -c test.conf sync 
----------------------------------------------------------------------------
ERROR: cmd_preexec "/usr/bin/sleep infinity" returned 139
```


Typically, the core dump flag is not included in the return code. Instead, the return code is normalized by ignoring the core dump flag and returning 128 plus the signal number in all cases. That is the reason I have the following return:
```perl
return (128 + ($retval & 0x7f));
```

DrHyde reviewed (changes requested):
I agree, this does indeed need a bit of fixin'

sgpinkus commented:
I think the fix has to be a bit more involved. Currently `get_retval` completely ignores the terminated-by-signal case. I don't think there is a good fix where `get_retval` continues to return just an int. If it is going to return an int, I think the best thing it could do is nothing at all, but that is not entirely backwards compatible, and results in hard to interpret error codes (albeit retaining all information). Probably best would be to return a triple: `<code, was_signal, core_dumped>` but then all callers need to be updated to account for this too.

nzalev commented:
I can change the return of `get_retval` fairly easily, since it is only called in a handful of places.

However, returning just an int is a normal thing to do. Higher value return codes being used to indicate signal numbers is an expected behaviour, since that's what shells already do.

But suppose we do return `<code, signal, core_dumped>`. What would rnsapshot do with that information? There are really only two situations where termination by signal occurs: 1) the program was terminated by the user or 2) the program misbehaved in such a way that it was terminated by the OS. In both cases, it is a fatal condition and rsnapshot should print an error exit - but that's what it already does. Whether or not a core dump would have been produced is essentially irrelevant to rsnapshot, and probably irrelevant to the vast majority of users as well.

Really, the only issue right now is that rsnapshot prints a floating point where it should be printing an integer.

```
ERROR: /usr/local/bin/rsync returned 0.54296875 while processing root@my.host.net:/
```

The point of this error message is to print the return code. If rsync was run interactively and was terminated by SIGSEGV, the vast majority of shells would simply set the return code to 128 + signo (139 as I have in my examples above). I don't really see a reason why should rsnapshot set the return code differently.

Of course, a new error message could be introduced to explicitly say that it was terminated by a signal, but I find that a bit redundant since returning 128 plus the signal number is already the most common way to indicate termination by signal.

sgpinkus commented:
> However, returning just an int is a normal thing to do. 

It is? What does normal mean? I know bash (and probably other shells) do this. Is this in a standard somewhere? Links / refs would be helpful.

> But suppose we do return <code, signal, core_dumped>. What would rnsapshot do with that information?

Print an error message like this example [here](https://perldoc.perl.org/functions/system).

> Of course, a new error message could be introduced to explicitly say that it was terminated by a signal, but I find that a bit redundant since returning 128 plus the signal number is already the most common way to indicate termination by signal.

Yeah your probably right. It's definitely an improvement over a float! And it can't really break compatibility since it's already broken. I think we just will accept the 128+signal version when error fixed.

nzalev commented:
I have fixed the if condition and switched ` / 256` to ` >> 8` at the regular return.

Both are looking good:

Normal return code (rsync is not installed at dest - return code = 12 for protocol error):
```
./rsnapshot -c test.conf sync
(admin@gw) Password for admin@gw1.home.local:
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
./rsnapshot -c test.conf sync 
----------------------------------------------------------------------------
ERROR: /usr/bin/rsync returned 12 while processing admin@gw:/
```

Killed by signal:
```
./rsnapshot -c test.conf sync
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
./rsnapshot -c test.conf sync 
----------------------------------------------------------------------------
ERROR: cmd_preexec "/usr/bin/sleep infinity" returned 139
```


---

I can work on changing up the return of `get_retval` and its callers to indicate if a termination by signal occurred and add a new error message at some point.

sgpinkus reviewed (commented):
:+1: Looks good. Thanks.

sgpinkus commented:
Oops, sorry I thought I approved merge already. Merging. Thanks @nzalev.
